File: backend/app/api/schemas/problem.py
from typing import List, Optional
from pydantic import BaseModel
from datetime import datetime

# Base Problem Schema
class ProblemBase(BaseModel):
    title: str
    statement: str
    difficulty: int  # 1-9
    topics: List[str]
    is_published: bool = False

# Schema for creating a problem
class ProblemCreate(ProblemBase):
    pass

# Schema for updating a problem
class ProblemUpdate(BaseModel):
    title: Optional[str] = None
    statement: Optional[str] = None
    difficulty: Optional[int] = None
    topics: Optional[List[str]] = None
    is_published: Optional[bool] = None

# Schema for problem response
class ProblemResponse(ProblemBase):
    id: int
    created_by: int
    created_at: datetime

    class Config:
        orm_mode = True 

File: frontend/src/types/problem.ts
import { User } from './user'; // Import the User type

// Base properties shared by create/update/read
interface ProblemBase {
    title: string;
    statement: string;
    difficulty: number;
    topics?: string[] | null;
    is_published: boolean;
}

// Type for creating a problem (matches ProblemCreate schema)
export interface ProblemCreate extends ProblemBase {}

// Type for updating a problem (matches ProblemUpdate schema)
export interface ProblemUpdate {
    title?: string;
    statement?: string;
    difficulty?: number;
    topics?: string[] | null;
    is_published?: boolean;
}

// Type for reading a problem (matches Problem schema)
export interface Problem extends ProblemBase {
    id: number;
    created_by_id: number;
    created_at: string; // Represent datetimes as strings
    creator: User; // Nested creator details
} 